File: 637.java
// 637. Average of Levels in Binary Tree
//   Given a non-empty binary tree, return the average value of the nodes on each level in the form of an array.
//   Example 1:
//
//   Input:
//       3
//      / \
//     9  20
//       /  \
//      15   7
//   Output: [3, 14.5, 11]
//   Explanation:
//   The average value of nodes on level 0 is 3,  on level 1 is 14.5, and on level 2 is 11. Hence return [3, 14.5, 11].
//   Note:
//
//   The range of node's value is in the range of 32-bit signed integer.
//
// Runtime: 2 ms, faster than 83.97% of Java online submissions for Average of Levels in Binary Tree.
// Memory Usage: 40.9 MB, less than 6.48% of Java online submissions for Average of Levels in Binary Tree.
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<Double> result = new ArrayList<>();
        queue.offer(root);
        while (!queue.isEmpty()) {
            int n = queue.size();
            Double sum = 0.0;
            for (int i = 0; i < n; i++) {
                TreeNode cur = queue.poll();
                sum = sum + Double.valueOf(cur.val);
                if (cur.left != null) {
                    queue.offer(cur.left);
                }
                if (cur.right != null) {
                    queue.offer(cur.right);
                }  
            }
            result.add(sum / n);
        }
        return result;
    }
}

File: go/637.go
// 637. Average of Levels in Binary Tree

// Given the root of a binary tree, return the average value of the nodes on each level in the form of an array. Answers within 10-5 of the actual answer will be accepted.
 

// Example 1:


// Input: root = [3,9,20,null,null,15,7]
// Output: [3.00000,14.50000,11.00000]
// Explanation: The average value of nodes on level 0 is 3, on level 1 is 14.5, and on level 2 is 11.
// Hence return [3, 14.5, 11].
// Example 2:


// Input: root = [3,9,20,15,7]
// Output: [3.00000,14.50000,11.00000]
 

// Constraints:

// The number of nodes in the tree is in the range [1, 104].
// -231 <= Node.val <= 231 - 1

// Runtime 0 ms Beats 100.00%
// Memory 8.34 MB Beats 56.94%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func averageOfLevels(root *TreeNode) []float64 {
    output := []float64{}
    queue := []*TreeNode{}
    queue = append(queue, root)
    for ; len(queue) > 0; {
        size := len(queue)
        sum := 0
        for i := 0; i < size; i++ {
            cur := queue[i]
            sum += cur.Val
            if cur.Left != nil {
                queue = append(queue, cur.Left)
            }
            if cur.Right != nil {
                queue = append(queue, cur.Right)
            }
        }
        output = append(output, float64(sum) / float64(size))
        queue = queue[size:]
    }
    return output
}